File: algorithms/searching/linear-search/implementation.js
/** Linear Search - O(n) time */

function linearSearch(arr, target) {
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] === target) return i;
    }
    return -1;
}

function linearSearchAll(arr, target) {
    const indices = [];
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] === target) indices.push(i);
    }
    return indices;
}

if (require.main === module) {
    console.log('Find 7:', linearSearch([4, 2, 7, 1, 9], 7));  // 2
    console.log('Find all 3s:', linearSearchAll([3, 1, 3, 7, 3], 3));  // [0,2,4]
}

module.exports = { linearSearch, linearSearchAll };

File: algorithms/searching/linear-search/implementation.py
"""Linear Search - O(n)"""

def linear_search(arr, target):
    """
    >>> linear_search([4, 2, 7, 1, 9], 7)
    2
    >>> linear_search([1, 2, 3], 5)
    -1
    >>> linear_search([], 1)
    -1
    """
    for i, val in enumerate(arr):
        if val == target:
            return i
    return -1

if __name__ == '__main__':
    import doctest
    r = doctest.testmod()
    print(f'✓ {r.attempted} tests passed!' if r.failed == 0 else f'✗ {r.failed} failed')
